src/assembler/rvassembler_common.h

```diff
@@ -6,49 +6,208 @@
 namespace Ripes {
 namespace Assembler {
 
-// The following macros assumes that ASSEMBLER_TYPES(..., ...) has been defined
-// for the given assembler.
+// A base class for RISC-V instructions
+template <typename Reg_T>
+class RVInstruction : public Instruction<Reg_T> {
+public:
+  RVInstruction(const Opcode<Reg_T> &opcode,
+                const std::vector<std::shared_ptr<Field<Reg_T>>> &fields)
+      : Instruction<Reg_T>(opcode, fields) {}
+};
 
-#define BType(name, funct3)                                                    \
-  std::shared_ptr<_Instruction>(new _Instruction(                              \
-      _Opcode(name,                                                            \
-              {OpPart(RVISA::Opcode::BRANCH, 0, 6), OpPart(funct3, 12, 14)}),  \
-      {std::make_shared<_Reg>(isa, 1, 15, 19, "rs1"),                          \
-       std::make_shared<_Reg>(isa, 2, 20, 24, "rs2"),                          \
-       std::make_shared<_Imm>(                                                 \
-           3, 13, _Imm::Repr::Signed,                                          \
-           std::vector{ImmPart(12, 31, 31), ImmPart(11, 7, 7),                 \
-                       ImmPart(5, 25, 30), ImmPart(1, 8, 11)},                 \
-           _Imm::SymbolType::Relative)}))
+class RVOpPartOpcode : public OpPart {
+public:
+  RVOpPartOpcode(RVISA::Opcode opcode)
+      : OpPart(opcode, 0, 6) {}
```

Comment:
Would be nice with a little note for all of these with the style of "All RISC-V opcodes are defined as the 7 LSBs of the instruction" or similar.